File: Engine/iOS/ParityCapture.swift
//
//  ParityCapture.swift — the NATIVE CAPTURE seam of the parity contract on iOS
//  (OpenSource/Conformance/parity/README.md, "The native capture planes"; the
//  Kotlin twin is Engine/Android render/ParityCapture.kt). The record lane's
//  specimen pass (ClosedSource/scripts/conformance/ParityRecord.swift) arms a
//  Session around a fixture render; StackNodeView's content chain consults it once
//  per element render, so every authored element that materializes reports its
//  settled geometry + the style facts the resolver produced for it. Disarmed
//  (session == nil — every production render), the cost is one static optional
//  read per element and nothing attaches.
//
//  StackNode is a VALUE type here (unlike the Kotlin twin's identity map), so the
//  comparable set is carried by a STAMP: ParityCapture.stamp walks the parsed
//  fixture tree and writes each authored node's source pre-order path into a
//  reserved attribute (`data-dsx-parity`). Only the record harness ever stamps, so
//  only the fixture's own authored nodes report — component internals and synthetic
//  nodes fall out naturally (platform anatomy, out of the comparable set by the
//  README's alignment rule). A node rendered more than once (bound list/grid rows)
//  records one entry per probe instance; the emitter orders repeats by settled
//  position and suffixes the path (`0.3.0~1`), exactly like the Kotlin twin.
//

import SwiftUI
import UIKit

enum ParityCapture {
    /// The reserved stamp attribute. Never authored; written only by `stamp`.
    static let pathKey = "data-dsx-parity"

    /// Head/declaration tags: registered, never boxed — the web plane has no host
    /// node for them either (the Kotlin twin's declarationTags).
    static let declarationTags: Set<String> = [
        "head", "event", "expects", "action", "api", "variable", "var", "let",
        "component", "formula", "script", "functions", "style", "watch", "attribute",
        "slot",
    ]

    /// The armed session, set ONLY by the record harness around a fixture render
    /// (main-thread renders; the harness arms, renders, settles, disarms).
    static var session: Session? = nil

    /// Stamp the authored tree with source pre-order paths. Child INDICES count every
    /// child (declaration siblings included) so paths agree with the Kotlin twin's
    /// indexing; declaration tags themselves are never stamped (they never box).
    static func stamp(_ node: StackNode, path: String = "0") -> StackNode {
        var copy = node
        if !declarationTags.contains(node.tag) {
            copy.attrs[pathKey] = path
        }
        copy.children = node.children.enumerated().map { index, child in
            stamp(child, path: "\(path).\(index)")
        }
        return copy
    }

    /// One probe instance's settled report: geometry in window points (CSS px 1:1)
    /// plus the RESOLVED attr strings the cascade produced (colors resolve to
    /// concrete rgb at emit, under the pass's committed scheme).
    struct Record {
        let path: String
        let tag: String
        let frame: CGRect
        let radius: Double
        let color: String?
        let background: String?
        let fontSize: Double?
        let fontWeight: String?
        let lineHeight: Double?
    }

    /// One measured node of the emitted plane — the web NodeMetrics twin.
    struct Measured {
        let path: String
        let tag: String
        let box: [Double]
        let radius: String
        let color: String
        let background: String
        let text: (size: String?, weight: String?, line: String?)?
    }

    final class Session {
        private var records: [UUID: Record] = [:]

        func report(slot: UUID, path: String, tag: String, frame: CGRect, attrs: [String: String]) {
            let fontSize = tag == "text" ? attrs["fontSize"].flatMap { Double($0) } : nil
            let lineHeight = fontSize.flatMap { fs in attrs["lineSpacing"].flatMap { Double($0) }.map { fs + $0 } }
            records[slot] = Record(
                path: path, tag: tag, frame: frame,
                radius: attrs["radius"].flatMap { Double($0) } ?? 0,
                color: attrs["color"],
                background: attrs["background"],
                fontSize: fontSize,
                fontWeight: tag == "text" ? attrs["fontWeight"] : nil,
                lineHeight: lineHeight
            )
        }

        var count: Int { records.count }

        /// Collapse the recorded probes into the ordered node list: source pre-order,
        /// repeats ordered by (y, x) with a `~k` appearance suffix; colors resolved
        /// under `style` (the pass's committed scheme).
        func measured(style: UIUserInterfaceStyle) -> [Measured] {
            var byPath: [String: [Record]] = [:]
            for r in records.values { byPath[r.path, default: []].append(r) }
            var out: [Measured] = []
            for (path, unordered) in byPath {
                let list = unordered.sorted { a, b in
                    a.frame.minY != b.frame.minY ? a.frame.minY < b.frame.minY : a.frame.minX < b.frame.minX
                }
                // A probe that re-attached in place reports one duplicate box — keep
                // distinct boxes only (real repeats occupy distinct positions).
                var distinct: [Record] = []
                for r in list where !distinct.contains(where: { $0.frame == r.frame }) { distinct.append(r) }
                for (k, r) in distinct.enumerated() {
                    let text: (size: String?, weight: String?, line: String?)? = r.tag == "text"
                        ? (r.fontSize.map(ParityCapture.formatPx), r.fontWeight, r.lineHeight.map(ParityCapture.formatPx))
                        : nil
                    out.append(Measured(
                        path: k == 0 ? path : "\(path)~\(k)",
                        tag: r.tag,
                        box: [round2(r.frame.minX), round2(r.frame.minY), round2(r.frame.width), round2(r.frame.height)],
                        radius: ParityCapture.formatPx(r.radius),
                        color: ParityCapture.cssColor(r.color ?? "label", style: style),
                        background: r.background.map { ParityCapture.cssColor($0, style: style) } ?? "rgba(0, 0, 0, 0)",
                        text: text
                    ))
                }
            }
            out.sort { ParityCapture.pathOrdered($0.path, $1.path) }
            return out
        }

        private func round2(_ v: CGFloat) -> Double { (Double(v) * 100).rounded() / 100 }
    }

    /// The capture probe — attached by StackNodeView's content chain OUTSIDE the
    /// styled view, so the reported bounds are the element's full styled box (the
    /// getBoundingClientRect / onGloballyPositioned twin). GeometryReader content
    /// evaluates at layout time, so the reported frame is the settled one; the
    /// harness additionally requires two consecutive identical snapshots.
    struct Probe: View {
        let session: Session
        let path: String
        let tag: String
        let attrs: [String: String]
        @State private var slot = UUID()

        init(session: Session, path: String, tag: String, attrs: [String: String]) {
            self.session = session
            self.path = path
            self.tag = tag
            self.attrs = attrs
        }

        var body: some View {
            GeometryReader { proxy -> Color in
                session.report(slot: slot, path: path, tag: tag,
                               frame: proxy.frame(in: .global), attrs: attrs)
                return Color.clear
            }
        }
    }

    /// Numeric source-path order: `0.10` sorts after `0.2`, `~k` repeats after the base.
    static func pathOrdered(_ a: String, _ b: String) -> Bool {
        func parts(_ p: String) -> ([Int], Int) {
            let split = p.split(separator: "~", maxSplits: 1)
            let base = split[0].split(separator: ".").map { Int($0) ?? 0 }
            let repeatIndex = split.count > 1 ? (Int(split[1]) ?? 0) : 0
            return (base, repeatIndex)
        }
        let (pa, ka) = parts(a)
        let (pb, kb) = parts(b)
        for i in 0..<max(pa.count, pb.count) {
            guard i < pa.count else { return true }
            guard i < pb.count else { return false }
            if pa[i] != pb[i] { return pa[i] < pb[i] }
        }
        return ka < kb
    }

    static func formatPx(_ v: Double) -> String {
        let rounded = (v * 100).rounded() / 100
        return rounded == rounded.rounded(.down) && rounded.isFinite
            ? "\(Int(rounded))px"
            : "\(rounded)px"
    }

    /// Resolve a DSX color word/literal to concrete `rgb()`/`rgba()` under a scheme —
    /// the style engine's own resolver (StackStyle.color), resolved through UIKit's
    /// dynamic-color machinery so system slots answer per scheme.
    static func cssColor(_ raw: String, style: UIUserInterfaceStyle) -> String {
        let resolved = UIColor(StackStyle.color(raw))
            .resolvedColor(with: UITraitCollection(userInterfaceStyle: style))
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        resolved.getRed(&r, green: &g, blue: &b, alpha: &a)
        func ch(_ v: CGFloat) -> Int { max(0, min(255, Int((v * 255).rounded()))) }
        if a >= 0.999 { return "rgb(\(ch(r)), \(ch(g)), \(ch(b)))" }
        let alpha = (Double(a) * 100).rounded() / 100
        return "rgba(\(ch(r)), \(ch(g)), \(ch(b)), \(alpha))"
    }
}

File: Engine/Android/render/src/main/kotlin/despia/engine/render/ParityCapture.kt
//
//  ParityCapture.kt — the NATIVE CAPTURE seam of the parity contract
//  (OpenSource/Conformance/parity/README.md, "The native capture planes"). The
//  instrumentation harness (ParityCaptureInstrumentedTest) arms a Session around a
//  fixture render; `Modifier.decorate` consults it once per attached element chain, so
//  every authored element that materializes reports its settled geometry + the style
//  facts the resolver produced for it. Disarmed (session == null — every production
//  render), the cost is one @Volatile null read per element and nothing composes.
//
//  The capture is keyed by SOURCE IDENTITY: a Session indexes the parsed fixture tree
//  by object identity, so only the fixture's own authored nodes record (component
//  internals and synthetic nodes fall out naturally — they are platform anatomy, out of
//  the comparable set by the README's alignment rule). A node that renders more than
//  once (list/grid rows re-render one template per row) records one entry per attached
//  chain; the emitter orders repeats by position and suffixes the path (`0.3.0~1`).
//

package despia.engine.render

import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.layout.positionInRoot
import despia.engine.StackNode
import java.util.IdentityHashMap
import java.util.concurrent.ConcurrentHashMap

object ParityCapture {

    /** Head/declaration tags: registered, never boxed — the web plane has no host node
     * for them either. */
    val declarationTags: Set<String> = setOf(
        "head", "event", "expects", "action", "api", "variable", "var", "let",
        "component", "formula", "script", "functions", "style", "watch", "attribute",
        "slot",
    )

    /** The armed session, set ONLY by the capture harness around a fixture render. */
    @Volatile var session: Session? = null

    /** One attached element chain's settled report. Geometry is window px (translated
     * to plane CSS px at emit); style facts resolved at composition time, under the
     * pass's committed scheme. */
    class Record(
        val node: StackNode,
        val x: Float, val y: Float, val w: Float, val h: Float,
        val radius: Double,
        val color: Color, val background: Color,
        val fontSize: Double?, val fontWeight: String?, val lineHeight: Double?,
    )

    /** One measured node of the emitted plane — the web NodeMetrics twin, colors kept
     * as resolved Compose colors so the harness can pair a light and a dark pass. */
    class Measured(
        val path: String,
        val tag: String,
        val box: DoubleArray,
        val radius: String,
        val color: Color,
        val background: Color,
        val text: TextMetrics?,
    )

    class TextMetrics(val size: String?, val weight: String?, val line: String?)

    class Session(root: StackNode) {
        private val paths = IdentityHashMap<StackNode, String>()
        private val records = ConcurrentHashMap<Any, Record>()

        init {
            fun index(node: StackNode, path: String) {
                paths[node] = path
                node.children.forEachIndexed { i, child ->
                    index(child, "$path.$i")
                }
            }
            index(root, "0")
        }

        internal fun capturesNode(node: StackNode): Boolean =
            node.tag !in declarationTags && paths.containsKey(node)

        /** The capture arm of the element chain — leftmost in decorate, so the reported
         * bounds are the element's full styled box (the web getBoundingClientRect twin,
         * modulo the gap-ledger rows). Colors resolve HERE, inside composition, so the
         * pass's committed scheme (StackTheme.WithCompositionScheme) answers exactly
         * what the style engine painted. `attrs` is the RESOLVED map (the full DSX-CSS
         * cascade already ran), so radius/font facts are the resolver's own numbers. */
        internal fun modifier(node: StackNode, attrs: Map<String, String>): Modifier =
            Modifier.composed {
                val slot = remember { Any() }
                val background = attrs["background"]?.let { StackStyle.color(it) } ?: Color.Transparent
                val color = attrs["color"]?.let { StackStyle.color(it) } ?: StackStyle.color("label")
                val radius = attrs["radius"]?.toDoubleOrNull() ?: 0.0
                val fontSize = if (node.tag == "text") attrs["fontSize"]?.toDoubleOrNull() else null
                val fontWeight = if (node.tag == "text") attrs["fontWeight"] else null
                val lineHeight = fontSize?.let { fs -> attrs["lineSpacing"]?.toDoubleOrNull()?.let { fs + it } }
                Modifier.onGloballyPositioned { coords ->
                    val p = coords.positionInRoot()
                    records[slot] = Record(
                        node, p.x, p.y,
                        coords.size.width.toFloat(), coords.size.height.toFloat(),
                        radius, color, background, fontSize, fontWeight, lineHeight,
                    )
                }
            }

        /** Collapse the recorded chains into the ordered node list: source pre-order,
         * repeats ordered by (y, x) with a `~k` appearance suffix. `origin` and
         * `density` translate window px into the plane's CSS-px coordinates. */
        fun measured(origin: Offset, density: Float): List<Measured> {
            fun cssPx(v: Float): Double = Math.round(v.toDouble() / density * 100.0) / 100.0
            val byNode = HashMap<StackNode, MutableList<Record>>()
            for (r in records.values) byNode.getOrPut(r.node) { mutableListOf() }.add(r)
            val out = ArrayList<Measured>()
            for ((node, list) in byNode) {
                val path = paths[node] ?: continue
                list.sortWith(compareBy({ it.y }, { it.x }))
                // An element whose chain attached twice reports one duplicate box —
                // keep distinct boxes only (real repeats occupy distinct positions).
                val distinct = list.distinctBy { listOf(it.x, it.y, it.w, it.h) }
                distinct.forEachIndexed { k, r ->
                    val text = if (node.tag == "text") TextMetrics(
                        r.fontSize?.let { formatPx(it) },
                        r.fontWeight,
                        r.lineHeight?.let { formatPx(it) },
                    ) else null
                    out.add(Measured(
                        if (k == 0) path else "$path~$k",
                        node.tag,
                        doubleArrayOf(cssPx(r.x - origin.x), cssPx(r.y - origin.y), cssPx(r.w), cssPx(r.h)),
                        formatPx(r.radius),
                        r.color, r.background,
                        text,
                    ))
                }
            }
            out.sortWith(compareBy(PATH_ORDER) { it.path })
            return out
        }
    }

    /** Numeric source-path order: `0.10` sorts after `0.2`, `~k` repeats after the base. */
    val PATH_ORDER: Comparator<String> = Comparator { a, b ->
        fun parts(p: String): Pair<List<Int>, Int> {
            val base = p.substringBefore('~')
            val repeat = p.substringAfter('~', "").toIntOrNull() ?: 0
            return base.split('.').map { it.toIntOrNull() ?: 0 } to repeat
        }
        val (pa, ka) = parts(a)
        val (pb, kb) = parts(b)
        for (i in 0 until maxOf(pa.size, pb.size)) {
            val va = pa.getOrNull(i) ?: return@Comparator -1
            val vb = pb.getOrNull(i) ?: return@Comparator 1
            if (va != vb) return@Comparator va.compareTo(vb)
        }
        ka.compareTo(kb)
    }

    fun formatPx(v: Double): String {
        val rounded = Math.round(v * 100.0) / 100.0
        return if (rounded == Math.floor(rounded)) "${rounded.toLong()}px" else "${rounded}px"
    }

    fun cssColor(c: Color): String {
        fun ch(v: Float): Int = Math.round(v * 255f).coerceIn(0, 255)
        val r = ch(c.red); val g = ch(c.green); val b = ch(c.blue)
        return if (c.alpha >= 0.999f) "rgb($r, $g, $b)"
        else "rgba($r, $g, $b, ${Math.round(c.alpha * 100f) / 100f})"
    }
}
